### PR TITLE
add copilot seat analysis supports

### PR DESCRIPTION
**Pull request by DevOps-zhuang**
add support for copilot seat display, since it is a popular request.

1) Call copilot seats API by calling https://docs.github.com/en/rest/copilot/copilot-user-management?apiVersion=2022-11-28#list-all-copilot-seat-assignments-for-an-organization
2) Display the seats API result in API Response Table. under the original usage API area.
3) Display the seats assigned/never used seats/no use in last 7 days seats in seat analysis tab
image

Limit: since the seat API currently only support organization level query now. so it is suggested to config the organization name in the .env file even the scope is set to enterprise. if there is no organization name set in .env. 
the assigned seat is displayed to 0.

Additions to the README:

* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R73-R86): Added a new `Seat Analysis` section which provides a detailed explanation of the metrics related to seat usage.

New Vue components:

* [`src/components/SeatsAnalysisViewer.vue`](diffhunk://#diff-ab0ce56718bd87d55c7b6acfbd15c636826f0bdcd0fa3a836d950a609736f09cR1-R186): Created a new Vue component to display the seat analysis. The component shows the total number of assigned seats, the number of seats assigned but never used, and the number of seats with no activity in the last 7 days. It also displays a table of seats with no activity in the last 7 days, including those that have never been used.

Changes to existing Vue components:

* [`src/components/MainComponent.vue`](diffhunk://#diff-e9eea4ec690e7649e2f812dfa7620e6200d892450dceb379d53e72f9be2894afL34-R35): Modified the component to include the `seat analysis` in the tab items and to fetch seat data from the API. The seat data is then passed as a prop to the `SeatsAnalysisViewer` and `ApiResponse` components. [[1]](diffhunk://#diff-e9eea4ec690e7649e2f812dfa7620e6200d892450dceb379d53e72f9be2894afL34-R35) [[2]](diffhunk://#diff-e9eea4ec690e7649e2f812dfa7620e6200d892450dceb379d53e72f9be2894afR48-R56) [[3]](diffhunk://#diff-e9eea4ec690e7649e2f812dfa7620e6200d892450dceb379d53e72f9be2894afR65) [[4]](diffhunk://#diff-e9eea4ec690e7649e2f812dfa7620e6200d892450dceb379d53e72f9be2894afL81-R86) [[5]](diffhunk://#diff-e9eea4ec690e7649e2f812dfa7620e6200d892450dceb379d53e72f9be2894afL93-R99) [[6]](diffhunk://#diff-e9eea4ec690e7649e2f812dfa7620e6200d892450dceb379d53e72f9be2894afL127-R163)
* [`src/components/ApiResponse.vue`](diffhunk://#diff-0b3857f49e9bfcf881b2f87a8a87330f654e136b039ee996337362417973d6f6R14-R23): Modified the component to display the seat count when the `Show Assigned Seats count` button is clicked. [[1]](diffhunk://#diff-0b3857f49e9bfcf881b2f87a8a87330f654e136b039ee996337362417973d6f6R14-R23) [[2]](diffhunk://#diff-0b3857f49e9bfcf881b2f87a8a87330f654e136b039ee996337362417973d6f6R36-R45) [[3]](diffhunk://#diff-0b3857f49e9bfcf881b2f87a8a87330f654e136b039ee996337362417973d6f6R69-R81)

Additions to the API:

* [`src/api/ExtractSeats.ts`](diffhunk://#diff-72433de4a75976d16d921f968a55213c0628b3f14104deb8ff7db25bd61dd12aR1-R66): Created a new API function `getSeatsApi` to fetch seat data from the GitHub API. The function supports both real and mocked data.

New JSON files:

* [`src/assets/enterprise_response_sample_seats.json`](diffhunk://#diff-00fdb9f4d99517cd43371dee7fb7dfd307f4c11239afe022510e93b0c5e3eebaR1-R74): Added a new JSON file to store a sample response for seat data at the enterprise level.
* [`src/assets/organization_response_sample_seats.json`](diffhunk://#diff-00fdb9f4d99517cd43371dee7fb7dfd307f4c11239afe022510e93b0c5e3eebaR1-R74): Added a new JSON file to store a sample response for seat data at the organization level.

### CODE REVIEW

**Comment by jackyhuang1** [comment]
@martedesco can you review this?  Junqian is from Microsoft China team who has seen many customer metrics reporting requirement like the one he's suggesting.

File: src/components/SeatsAnalysisViewer.vue [hunk 1]
```diff
@@ -0,0 +1,186 @@
+<template>
+    <div class="tiles-container">      
+        <v-card elevation="4" color="white" variant="elevated" class="mx-auto my-3" style="width: 300px; height: 175px;">
+            <v-card-item>
+                <div>
+                <div class="text-overline mb-1" style="visibility: hidden;">filler</div>
+                <div class="text-h6 mb-1">Total Assigned  </div>
+                <div class="text-caption">
+                    currently assigned seats
```
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue]:
```suggestion
                    Currently assigned seats
```

File: src/components/SeatsAnalysisViewer.vue [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                <p>{{ NoshowSeats.length }}</p>
+                </div>
+            </v-card-item>
+        </v-card>
+        <v-card elevation="4" color="white" variant="elevated" class="mx-auto my-3" style="width: 300px; height: 175px;">
+            <v-card-item>
+                <div>
+                <div class="text-overline mb-1" style="visibility: hidden;">filler</div>
+                <div class="text-h6 mb-1">No Activity in Last 7 days </div>
+                <div class="text-caption">
```
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue:33]:
```suggestion
                <div class="text-h6 mb-1">No Activity in the Last 7 days</div>
```

File: src/components/SeatsAnalysisViewer.vue [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                    no use in last 7 days
```
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue]:
```suggestion
                    No use in the last 7 days
```

File: src/components/SeatsAnalysisViewer.vue [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                        { title: 'ID', key: 'id' },
+                        { title: 'Assigned Time', key: 'created_at' },
+                        { title: 'Last Activity At', key: 'last_activity_at' },
+                        { title: 'Last Activity Editor', key: 'last_activity_editor' },
+                    ]"
+                    :items="unusedSeats"
+                    :items-per-page="10"
+                    class="elevation-2"
+                >      
+                -->
```
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue]:
```suggestion   
```

File: src/components/SeatsAnalysisViewer.vue [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                <template v-slot:item="{ item }">
+                    <tr>
+                        <td>{{ item.login }}</td>
+                        <td>{{ item.id }}</td>
+                        <td>{{ item.created_at }}</td>
+                        <td>{{ item.last_activity_at }}</td>
+                        <td>{{ item.last_activity_editor }}</td>
+                    <!--
+                        <td>{{ item.last_activity_at }}</td>
+                    -->
```
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue]:
```suggestion
```

File: src/components/SeatsAnalysisViewer.vue [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+                </v-data-table>
+            </v-container>
+        </v-main>
+    </div>
+</template>
+  
+<script lang="ts">
+  import { defineComponent, ref, toRef , watchEffect } from 'vue';
+  import { Seat } from '../model/Seat';
+  import { Line, Bar } from 'vue-chartjs'
```
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue]:
```suggestion
```

File: src/components/SeatsAnalysisViewer.vue [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    },
+data() {
+    return {
+        headers: [
+            { title: 'Login', key: 'login' },
+            { title: 'GitHub ID', key: 'id' },
+            { title: 'Assigned to the Organization At', key: 'created_at' },
+            { title: 'Last Activity At', key: 'last_activity_at' },
+            { title: 'Last Activity Editor', key: 'last_activity_editor' },
+            // 添加其他需要显示的属性
```
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue]:
```suggestion
```

File: src/components/MainComponent.vue [hunk 8]
```diff
@@ -90,7 +95,8 @@ export default defineComponent({
   setup() {
       const metricsReady = ref(false);
       const metrics = ref<Metrics[]>([]);
-
+      const seatsReady = ref(false); // Add this line by zhuang to the setup function
+      const seats = ref<Seat[]>([]); // Add this line by zhuang to the setup function
```
Comment (martedesco) [inline src/components/MainComponent.vue]:
```suggestion
      const seatsReady = ref(false); 
      const seats = ref<Seat[]>([]); 
```

File: src/components/SeatsAnalysisViewer.vue (hunk 2 above)
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue:33]:
```suggestion
                <div class="text-h6 mb-1">No Activity in the Last 7 days</div>
```

File: README.md [hunk 9]
```diff
@@ -70,6 +70,20 @@ The language breakdown analysis tab also displays a table showing the Accepted P
...
 
+## Seat Analysis 
+
+![image](https://github.com/DevOps-zhuang/copilot-metrics-viewer/assets/54096296/d1fa9d1d-4fab-4e87-84ba-7be189dd4dd0)
+
+1. **Total Assigned:** This metric represents the total number of copilot assigned seats within current organization.
+
+2. **Assigned But Never Used:** This metric shows the assigned but never used seats within current organization. the assigned time are also dispalyed in the below chart, recommend to check that date.
+
+3. **No Activity in Last 7 days:** never used seats. or seats used, but no activity in recent 7 days.
```
Comment (martedesco) [inline README.md]:
```suggestion
3. **No Activity in the Last 7 days:** never used seats or seats used, but with no activity in the past 7 days.
```

File: README.md [hunk 10]
```diff
@@ -70,6 +70,20 @@ The language breakdown analysis tab also displays a table showing the Accepted P
 
 4. **Total Active Copilot Chat Users:** a bar chart that illustrates the total number of users who have actively interacted with Copilot over the past 28 days.
 
+## Seat Analysis 
+
+![image](https://github.com/DevOps-zhuang/copilot-metrics-viewer/assets/54096296/d1fa9d1d-4fab-4e87-84ba-7be189dd4dd0)
+
+1. **Total Assigned:** This metric represents the total number of copilot assigned seats within current organization.
```
Comment (martedesco) [inline README.md]:
```suggestion
1. **Total Assigned:** This metric represents the total number of Copilot seats assigned within current organization.
```

File: README.md [hunk 11]
```diff
@@ -70,6 +70,20 @@ The language breakdown analysis tab also displays a table showing the Accepted P
...
+
+2. **Assigned But Never Used:** This metric shows the assigned but never used seats within current organization. the assigned time are also dispalyed in the below chart, recommend to check that date.
```
Comment (martedesco) [inline README.md]:
```suggestion
2. **Assigned But Never Used:** This metric shows seats that were assigned but never within the current organization. The assigned timestamp is also displayed in the below chart.
```

File: README.md [hunk 12]
```diff
@@ -70,6 +70,20 @@ The language breakdown analysis tab also displays a table showing the Accepted P
...
+
+3. **No Activity in Last 7 days:** never used seats. or seats used, but no activity in recent 7 days.
+
+4. **No Activity in last 7 days (Inluding Never used seats):** A table to display seats that have had no activity in the past 7 days, ordered by the date of last activity. Seats that were used earlier are displayed more prominently at the front.
```
Comment (martedesco) [inline README.md]:
```suggestion
4. **No Activity in the last 7 days (including never used seats):**a table to display seats that have had no activity in the past 7 days, ordered by the date of last activity. Seats that were used earlier are displayed at the top.
```

File: src/components/ApiResponse.vue [hunk 13]
```diff
@@ -11,6 +11,16 @@
           <div v-if="showCopyMessage" :class="{'copy-message': true, 'error': isError}">{{ message }}</div>
         </transition>
       </div>
+      
+      <v-card max-height="575px" class="overflow-y-auto">
```
Comment (martedesco) [inline src/components/ApiResponse.vue:17]:
```suggestion
      <br><br>
      <v-card max-height="575px" class="overflow-y-auto">
```

Adding line breaks

File: src/components/ApiResponse.vue [hunk 14]
```diff
@@ -11,6 +11,16 @@
...
+          <pre ref="jsonText">{{ JSON.stringify(seats, null, 2) }}</pre>
+      </v-card>
```
Comment (martedesco) [inline src/components/ApiResponse.vue:19]:
```suggestion
      </v-card>
      <br>
```
Line break

File: src/components/SeatsAnalysisViewer.vue [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+                </div>
+                <p>{{ NoshowSeats.length }}</p>
+                </div>
+            </v-card-item>
+        </v-card>
+        <v-card elevation="4" color="white" variant="elevated" class="mx-auto my-3" style="width: 300px; height: 175px;">
+            <v-card-item>
+                <div>
+                <div class="text-overline mb-1" style="visibility: hidden;">filler</div>
+                <div class="text-h6 mb-1">No Activity in Last 7 days </div>
```
Comment (martedesco) [inline src/components/SeatsAnalysisViewer.vue]:
```suggestion
                <div class="text-h6 mb-1">No Activity in the Last 7 days </div>
```

File: README.md [hunk 16]
```diff
@@ -70,6 +70,20 @@ The language breakdown analysis tab also displays a table showing the Accepted P
 
 4. **Total Active Copilot Chat Users:** a bar chart that illustrates the total number of users who have actively interacted with Copilot over the past 28 days.
 
+## Seat Analysis 
+
+![image](https://github.com/DevOps-zhuang/copilot-metrics-viewer/assets/54096296/d1fa9d1d-4fab-4e87-84ba-7be189dd4dd0)
```
Comment (martedesco) [inline README.md:75]: We might need to update this image once the changes have been implemented 🙏🏼

**Comment by martedesco** [comment]
@DevOps-zhuang Thanks a lot for your contribution. I am suggesting some minor changes on the labels and code clean up. Do you mind also updating the image in the README once the changes have been accepted? 🙏🏼

**Comment by DevOps-zhuang** [comment]
> @DevOps-zhuang Thanks a lot for your contribution. I am suggesting some minor changes on the labels and code clean up. Do you mind also updating the image in the README once the changes have been accepted? 🙏🏼

Thanks, I will update the README. and I just add some information to state the limit of this feature, please also check it. and Since I don't have a copilot for business  enabled in GitHub Enterprise level. could you please help to check it when the scope is set to enterprise in .env file? thanks.

> Limit: since the seat API currently only supports organization level query now. so it is suggested to config the organization name in the .env file even the scope is set to enterprise. if there is no organization name found in .env.
 the assigned seat is displayed to 0.

**Comment by martedesco** [comment]
> Since I don't have a copilot for business enabled in GitHub Enterprise level. could you please help to check it when the scope is set to enterprise in .env file? thanks.

When the scope is set to `enterprise`, it will still display the results of the `organization` defined in the `.env` file. Take a look at the image below. 

image

How about we don't show the Seat Analysis tab when it is `enterprise` scope?

**Comment by martedesco** [comment]
> Thanks Marcio for your check and suggestion, I fully accepted your suggestion, and add some statement about the feature limit, the main limit comes for the current seat API only support organization level query. so when the scope is set as enterprise in .env, and there is no organization set, the default assigned seat will be displayed as 0.

We can mention that limitation in the README and in the app we can opt for not displaying the Seat Analysis tab when the scope is `enterprise`

File: src/components/MainComponent.vue [hunk 17]
```diff
@@ -31,7 +31,10 @@
               <BreakdownComponent v-if="item === 'languages'" :metrics="metrics" :breakdownKey="'language'"/>
               <BreakdownComponent v-if="item === 'editors'" :metrics="metrics" :breakdownKey="'editor'"/>
               <CopilotChatViewer v-if="item === 'copilot chat'" :metrics="metrics" />
-              <ApiResponse v-if="item === 'api response'" :metrics="metrics" />
+              <div v-if="isScopeOrganization">
+                <SeatsAnalysisViewer v-if="item === 'seat analysis'" :seats="seats" />
```
Comment (martedesco) [inline src/components/MainComponent.vue:35]: @DevOps-zhuang I added this condition so it will only display if in the organization scope

**Comment by DevOps-zhuang** [comment]
Thanks for update, you are so quick! I will check and study it later 😊

File: src/components/MainComponent.vue (hunk 17 above)
Comment (DevOps-zhuang) [reply src/components/MainComponent.vue:35]: great！ it is better than my code. where I set the v-if to the <SeatsAnalysisViewer>, and it always display the SeatsAnalysisViewer, thouth there is no content when the scopt is not organize, I will accept your code, thanks)

**Comment by DevOps-zhuang** [comment]
@martedesco I added two more commits for the conditional control. one in the ExtractSeats.ts to filter the Seat API only when scope is 'organization'; and display the 'seat' information in APIresponse within same condition; could you please review it? thanks)

File: src/components/ApiResponse.vue
Comment (DevOps-zhuang) [inline src/components/ApiResponse.vue:1]: @martedesco , I made some updates to only  display the 'seat information' when scope is 'organization', thanks for check!

File: src/components/MainComponent.vue [hunk 18]
```diff
@@ -5,7 +5,7 @@
         <v-icon>mdi-github</v-icon>
       </v-btn>
 
-      <v-toolbar-title>Copilot Metrics Viewer | {{ capitalizedItemName }} : {{ gitHubOrgName }}</v-toolbar-title>
```
Comment (DevOps-zhuang) [inline src/components/MainComponent.vue:8]: when the scope is enterprise, it is better to display the VUE_APP_GITHUB_ENTname here, not the gitHubOrgName

**Review by DevOps-zhuang — COMMENTED** [review]
just upate the view to make it more readable, thanks.

**Comment by martedesco** [comment]
@DevOps-zhuang It looks good to me :) I just fixed a small typo and I'm merging this one 👍🏼🚀

**Review by martedesco — APPROVED** [review]

File: README.md (hunk 16 above)
Comment (martedesco) [reply README.md:75]: Issue created: [README Seat Analysis image update](https://github.com/github-copilot-resources/copilot-metrics-viewer/issues/35)

**Review by martedesco — APPROVED** [review]

**Pull request merged**